### ISSUE TITLE
[Heartbeat] Write summary events on process exit 

### ISSUE THREAD

**Issue by vigneshshanmugam** [elastic/beats#31948]
From the discussion with @paulb-elastic, he brought up one interesting aspect of propagating errors from Heartbeat process. 

The HB instance running in the service can be killed by the platform process or by the user in a multitude of ways while performing a synthetics run. Some of the examples are 

1. The browser monitor is running way longer than expected and the HB pod is killed by the platform.
2. OOM issues with chrome and Synthetics process is hanging and eventually killed by the platform.  

When both of these happens during the middle of a monitor execution, The monitor run would not yield any summary event resulting in users not seeing any status for that run in the Kibana UI. 

### Solution

There is a bigger scope of solving the error propagation/reporting from the Synthetics service to the Kibana UI which is covered on this issue https://github.com/elastic/synthetics-service/issues/100. But as the HB already has the credentials to write back to the customer ES cluster, it would be easier to deal with the process timeout/exit issue in the HB side and have a valuable feedback for the monitor run.

**Comment by andrewvc** [comment]
@vigneshshanmugam that behavior already exists in https://github.com/elastic/beats/pull/31936 , LMK if you're thinking of something else here.

**Comment by vigneshshanmugam** [comment]
I dont think so Andrew, The code on the HB side can write proper summary docs when the Synthetics process exited, not when the HB process itself exits in the middle of a browser monitor run. I haven't confirmed it though, I can try it out first and post more details here.

**Comment by andrewvc** [comment]
Ah, I misunderstood, that makes sense. 

Based on [this description of the OOM Killer algorithm](https://unix.stackexchange.com/questions/153585/how-does-the-oom-killer-decide-which-process-to-kill-first), it seems like the kernel would always pick chrome to kill first. So, the OOM case seems unlikely.

The other case, of a too long journey, could easily be enforced by heartbeat. I think in general the best pattern is to have multiple execution limits. Heartbeat could easily have a execution time limit added. In that case users should never see the hard limit hit unless they're trying to circumvent that or extremely weird circumstances occur.

In other words, by adding that check at the heartbeat level we can probably achieve 90% of the benefit for 10% of the work. Thoughts?

**Comment by vigneshshanmugam** [comment]
There are two ways I can think of where this comes in to play
1. SIGTERM when we exceed the active deadline for a particular kubernetes job timeout - https://kubernetes.io/docs/concepts/workloads/controllers/job/#job-termination-and-cleanup:~:text=Another%20way%20to,reason%3A%20DeadlineExceeded.
2. OOM killer that you posted above. 

Agreed with you to have multiple execution limits. But I am just worried about would we pick as the correct limit? Instead of doing it on the HB side, Wouldn't it be simpler to add a handler on the process exit and leave the hard limit to the platform (k8s in our case).

**Comment by andrewvc** [comment]
As mentioned before, I don't believe the OOM case is one we should design for since I don't think it will affect the heartbeat process. So, in the case of an OOM users should get an OOM and things should basically work right. We probably should test this and ensure the error message we send to the stack is decent.

WRT timeouts, I think it's simpler to implement it in HB. It already has an ES connection and possibly partial state for the current check.  In addition to this, that functionality would be useful for self-managed instances, where we can't depend on K8s or other outside helpers.

**Comment by vigneshshanmugam** [comment]
Agreed that the functionality would be super helpful for the self-managed instances where the monitor check exceeds a specified timeout. Timeout issue is tracked here - #32388

**Comment by paulb-elastic** [comment]
Closing as https://github.com/elastic/beats/issues/32388 has been implemented part of this, and https://github.com/elastic/synthetics-dev/issues/129 is following up.

@vigneshshanmugam if something else missing, please open a separate issue
